pytket_qir/converter.py

```diff
@@ -0,0 +1,357 @@
+                self.local_conditions[block_name] = self.parser.ssa_vars[condition_name]
+            elif isinstance(term, (QirBrTerminator, QirRetTerminator)):
+                # Return is considered an unconditional branch.
+                self.local_conditions[block_name] = condition_bit
+            preds = block.preds
+            if not preds:  # Entry block.
+                expr = self.local_conditions["entry_pred"]
+                self.conditions[block_name] = self.conditions[block_name] | (
+                    self.conditions["entry_pred"] & expr
+                )
+            else:
+                # import pdb; pdb.set_trace()
+                for pred in preds:
+                    if (
+                        self.edges[pred][block_name] == None
+                        or self.edges[pred][block_name] == True
+                    ):
+                        expr = self.local_conditions[pred]
+                    elif self.edges[pred][block_name] == False:
+                        expr = BitNot(self.local_conditions[pred])
+                    self.conditions[block_name] = self.conditions[block_name] | (
+                        self.conditions[pred] & expr
+                    )
+
+            # Add extra bits created in the sub-circuit to the main one.
+            for bit in set(circuit.bits) - set(main_circuit.bits):
+                main_circuit.add_bit(bit)
+            circ_box = CircBox(circuit)
+            arguments = [qubit.index[0] for qubit in circuit.qubits] + [
+                bit.index[0] for bit in circuit.bits
+            ]  # Order matters.
+            main_circuit.add_circbox(
+                circ_box, arguments, condition=if_bit(self.conditions[block_name])
+            )
+        return main_circuit
+
+    def apply_contraction(self, block) -> Block:
+        term = block.terminator
+        has_jump = isinstance(term, QirBrTerminator)
+        curr_block = self.cfg[block.name]
+        curr_block.visited = True
+        if has_jump:
+            next_block = cast(
+                QirBlock, self.module.functions[0].get_block_by_name(term.dest)
+            )
+            # A precondition to apply the rewrite rule.
+            continue_next = len(self.cfg[next_block.name].preds) == 1
+
+            if continue_next:  # Recurse through the chain of jumps.
+                next_block_inst = self.apply_contraction(next_block)
+                return Block(
+                    name=curr_block.name,
+                    succs=next_block_inst.succs,
+                    preds=curr_block.preds,
+                    composition=[block.name] + next_block_inst.composition,
+                )
+            else:  # Return the current block.
+                return curr_block
+        return curr_block
+
+    def collapse_blocks(self) -> None:
+        """
+        Detect and collapse chains of linear blocks.
+
+        Return a rewritten CFG where all blocks have in-degree exactly one
```

Comment:
How are you taking care of the `in-degree exactly one` in the case where you are combining two branches to one?

Reply:
Good point, this is inexact. I've amended it.